File: 2025/02.py
import time
start_time = time.time()

with open("02.txt", "r") as file:
    lines = [line.rstrip() for line in file]  # can be just strip()

# only one line
line = lines[0]


def is_valid_id(n):
    id = str(n)

    # odd number of digits cant be invalid
    if (len(id) % 2 != 0):
        return False

    first_half = id[:len(id) // 2]
    second_half = id[len(id) // 2:]

    if (first_half == second_half):
        return True

    return False


def is_valid_id2(n):
    id = str(n)

    # build the numbers from the head from 1 to len/2 characters -> compare with original string
    for i in range(1, len(id) // 2 + 1):
        if (len(id) % i == 0):
            # build string
            # get head
            head = id[:i]

            # create string body
            body = ""
            for j in range(1, len(id) // i + 1):
                body += head

            if (id == body):
                return True

    return False


# part 1
ranges = line.split(',')
sum = 0

for r in ranges:
    split = r.split('-')

    start = int(split[0])
    stop = int(split[1]) + 1

    for i in range(start, stop):
        if (is_valid_id(i)):
            sum += i

print(f"Part 1: {sum}")


# part 2
ranges2 = line.split(',')
sum2 = 0

for r in ranges2:
    split = r.split('-')

    start = int(split[0])
    stop = int(split[1]) + 1

    for i in range(start, stop):
        if (is_valid_id2(i)):
            sum2 += i

print(f"Part 2: {sum2}")


end_time = time.time()
time_differnece = end_time - start_time
print(f"Elapsed time: {time_differnece} seconds")

File: 2025/02.lua
local start_time = os.clock()

local lines = {}
for line in io.lines("02.txt") do
	table.insert(lines, line)
end

-- only one line
local line = lines[1]


local function is_invalid_id(n)
	local id = tostring(n)

	-- odd number of digits cant be invalid
	if (#id%2 ~= 0) then
		return false
	end

	local first_half = string.sub(id, 1, #id/2)
	local second_half = string.sub(id, (#id/2)+1)

	if first_half == second_half then
		return true
	end

	return false
end

local function is_invalid_id2(n)
	local id = tostring(n)

	-- build the numbers from the head from 1 to len/2 characters -> compare with og string
	for i = 1, #id/2 do
		if #id%i == 0 then
			--build string
			-- get head
			local head = string.sub(id, 1, i)

			-- create string body
			local body = ""
			for _ = 1, #id/i do
				body = body .. head
			end

			if id == body then
				return true
			end

		end
	end

	return false
end


-- part 1
local ranges = {}
local sum = 0

for i in string.gmatch(line, "[^,]+") do -- split string by ,
	table.insert(ranges, i)
end

for i = 1, #ranges do
	local split = string.find(ranges[i], "-")
	if split==nil then
		break
	end
	local start = tonumber(string.sub(ranges[i], 1, split-1))
	local stop = tonumber(string.sub(ranges[i], split+1))
	-- tonumber() only here to make an int (otherwise it will be a flow in for loop)

	for j = start, stop do
		if is_invalid_id(j) then
			sum = sum + j
		end
	end
end

print("Part 1: " .. sum)


-- part 2
local ranges2 = {}
local sum2 = 0

for i in string.gmatch(line, "[^,]+") do -- split string by ,
	table.insert(ranges2, i)
end

for i = 1, #ranges2 do
	local split = string.find(ranges2[i], "-")
	if split==nil then
		break
	end
	local start = tonumber(string.sub(ranges2[i], 1, split-1))
	local stop = tonumber(string.sub(ranges2[i], split+1))

	for j = start, stop do
		if is_invalid_id2(j) then
			sum2 = sum2 + j
		end
	end
end

print("Part 2: " .. sum2)


local end_time = os.clock()
print("Elapsed time: " .. (end_time - start_time) ..  " seconds")
